File: objectModel/Python/cdm/persistence/cdmfolder/document_persistence.py
# Copyright (c) Microsoft Corporation. All rights reserved.
# Licensed under the MIT License. See License.txt in the project root for license information.

from typing import List

from cdm.enums import CdmObjectType
from cdm.persistence import PersistenceLayer
from cdm.objectmodel import CdmCorpusContext, CdmDocumentDefinition, CdmEntityDefinition
from cdm.utilities import CopyOptions, ResolveOptions, copy_data_utils, logger

from cdm.enums import CdmLogCode
from .attribute_group_persistence import AttributeGroupPersistence
from .constant_entity_persistence import ConstantEntityPersistence
from .data_type_persistence import DataTypePersistence
from .entity_persistence import EntityPersistence
from .import_persistence import ImportPersistence
from .purpose_persistence import PurposePersistence
from .trait_persistence import TraitPersistence
from .trait_group_persistence import TraitGroupPersistence
from .types import DocumentContent

_TAG = 'DocumentPersistence'

class DocumentPersistence:
    is_persistence_async = False

    formats = [PersistenceLayer.CDM_EXTENSION]

    # The maximum json semantic version supported by this ObjectModel version.
    json_semantic_version = CdmDocumentDefinition.current_json_schema_semantic_version

    @staticmethod
    def from_data(ctx: 'CdmCorpusContext', doc_name: str, json_data: str, folder: 'CdmFolderDefinition') -> 'CdmDocumentDefinition':
        obj = DocumentContent().decode(json_data)
        return DocumentPersistence.from_object(ctx, doc_name, folder.namespace, folder.folder_path, obj)

    @staticmethod
    def from_object(ctx: CdmCorpusContext, name: str, namespace: str, path: str, data: 'DocumentContent') -> 'CdmDocumentDefinition':
        document = ctx.corpus.make_object(CdmObjectType.DOCUMENT_DEF, name)
        document.folder_path = path
        document.namespace = namespace

        if data:
            if data.get('schema'):
                document.schema = data.schema

            # support old model syntax
            if data.get('schemaVersion'):
                document.json_schema_semantic_version = data.schemaVersion

            if data.get('documentVersion'):
                document.document_version = data.documentVersion

            if data.get('imports'):
                for import_obj in data.imports:
                    document.imports.append(ImportPersistence.from_data(ctx, import_obj))

            if data.get('definitions') and isinstance(data.definitions, List):
                for definition in data.definitions:
                    if definition.get('dataTypeName'):
                        document.definitions.append(DataTypePersistence.from_data(ctx, definition))
                    elif definition.get('purposeName'):
                        document.definitions.append(PurposePersistence.from_data(ctx, definition))
                    elif definition.get('attributeGroupName'):
                        document.definitions.append(AttributeGroupPersistence.from_data(ctx, definition))
                    elif definition.get('traitName'):
                        document.definitions.append(TraitPersistence.from_data(ctx, definition))
                    elif definition.get('traitGroupName'):
                        document.definitions.append(TraitGroupPersistence.from_data(ctx, definition))
                    elif definition.get('entityShape'):
                        document.definitions.append(ConstantEntityPersistence.from_data(ctx, definition))
                    elif definition.get('entityName'):
                        document.definitions.append(EntityPersistence.from_data(ctx, definition))

            is_resolved_doc = False
            if len(document.definitions) == 1 and isinstance(document.definitions[0], CdmEntityDefinition):
                entity = document.definitions[0]  # type: CdmEntityDefinition
                resolved_trait = entity.exhibits_traits.item('has.entitySchemaAbstractionLevel')
                # Tries to figure out if the document is in resolved form by looking for the schema abstraction trait
                # or the presence of the attribute context.
                is_resolved_doc = resolved_trait and resolved_trait.arguments[0].value == 'resolved'
                is_resolved_doc = is_resolved_doc or entity.attribute_context

            if data.jsonSchemaSemanticVersion:
                document.json_schema_semantic_version = data.jsonSchemaSemanticVersion
                if DocumentPersistence._compare_json_semantic_version(ctx, document.json_schema_semantic_version) > 0:
                    message = 'This ObjectModel version supports json semantic version {} at maximum.'.format(DocumentPersistence.json_semantic_version)
                    message += ' Trying to load a document with version {}.'.format(document.json_schema_semantic_version)
                    if is_resolved_doc:
                        logger.warning(ctx, _TAG, DocumentPersistence.from_data.__name__, None,
                                       CdmLogCode.WARN_PERSIST_UNSUPPORTED_JSON_SEM_VER, document.json_schema_semantic_version)
                    else:
                        logger.error(ctx, _TAG, DocumentPersistence.from_data.__name__, None,
                                     CdmLogCode.ERR_PERSIST_UNSUPPORTED_JSON_SEM_VER, document.json_schema_semantic_version)
            else:
                logger.warning(ctx, _TAG, DocumentPersistence.from_data.__name__, document.at_corpus_path, CdmLogCode.WARN_PERSIST_JSON_SEM_VER_MANDATORY)

        return document

    @staticmethod
    def to_data(instance: CdmDocumentDefinition, res_opt: ResolveOptions, options: CopyOptions) -> DocumentContent:
        result = DocumentContent()
        result.schema = instance.schema
        result.jsonSchemaSemanticVersion = instance.json_schema_semantic_version
        result.imports = copy_data_utils._array_copy_data(res_opt, instance.imports, options)
        result.definitions = copy_data_utils._array_copy_data(res_opt, instance.definitions, options)
        result.documentVersion = instance.document_version
        return result

    @staticmethod
    def _compare_json_semantic_version(ctx: 'CdmCorpusContext', document_semantic_version: str) -> int:
        """Compares the document version with the json semantic version supported.
        1 => if document_semantic_version > json_semantic_version
        0 => if document_semantic_version == json_semantic_version or if document_semantic_version is invalid
        -1 => if document_semantic_version < json_semantic_version"""
        curr_semantic_version_split = [int(x) for x in DocumentPersistence.json_semantic_version.split('.')]

        error_message = 'jsonSemanticVersion must be set using the format <major>.<minor>.<patch>.'

        try:
            doc_semantic_version_split = [int(x) for x in document_semantic_version.split('.')]
        except ValueError:
            logger.warning(ctx, _TAG, DocumentPersistence._compare_json_semantic_version.__name__, None,
                           CdmLogCode.WARN_PERSIST_JSON_SEM_VER_INVALID_FORMAT)
            return 0

        if len(doc_semantic_version_split) != 3:
            logger.warning(ctx, _TAG, DocumentPersistence._compare_json_semantic_version.__name__, None,
                           CdmLogCode.WARN_PERSIST_JSON_SEM_VER_INVALID_FORMAT)
            return 0

        for i in range(3):
            if doc_semantic_version_split[i] != curr_semantic_version_split[i]:
                return -1 if doc_semantic_version_split[i] < curr_semantic_version_split[i] else 1
        return 0

File: objectModel/Java/objectmodel/src/main/java/com/microsoft/commondatamodel/objectmodel/persistence/cdmfolder/DocumentPersistence.java
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License. See License.txt in the project root for license information.

package com.microsoft.commondatamodel.objectmodel.persistence.cdmfolder;

import com.fasterxml.jackson.databind.JsonNode;
import com.google.common.base.Strings;
import com.microsoft.commondatamodel.objectmodel.cdm.CdmCollection;
import com.microsoft.commondatamodel.objectmodel.cdm.CdmCorpusContext;
import com.microsoft.commondatamodel.objectmodel.cdm.CdmDocumentDefinition;
import com.microsoft.commondatamodel.objectmodel.cdm.CdmEntityDefinition;
import com.microsoft.commondatamodel.objectmodel.cdm.CdmFolderDefinition;
import com.microsoft.commondatamodel.objectmodel.cdm.CdmObjectDefinition;
import com.microsoft.commondatamodel.objectmodel.cdm.CdmTraitReference;
import com.microsoft.commondatamodel.objectmodel.persistence.CdmConstants;
import com.microsoft.commondatamodel.objectmodel.enums.CdmLogCode;
import com.microsoft.commondatamodel.objectmodel.enums.CdmObjectType;
import com.microsoft.commondatamodel.objectmodel.persistence.cdmfolder.types.DataType;
import com.microsoft.commondatamodel.objectmodel.persistence.cdmfolder.types.DocumentContent;
import com.microsoft.commondatamodel.objectmodel.persistence.cdmfolder.types.Import;
import com.microsoft.commondatamodel.objectmodel.utilities.CopyOptions;
import com.microsoft.commondatamodel.objectmodel.utilities.JMapper;
import com.microsoft.commondatamodel.objectmodel.utilities.ResolveOptions;
import com.microsoft.commondatamodel.objectmodel.utilities.logger.Logger;

public class DocumentPersistence {
  private static String tag = DocumentPersistence.class.getSimpleName();

  /**
   * Whether this persistence class has async methods.
   */
  public static final boolean isPersistenceAsync = false;
  
  /**
   * The file format/extension types this persistence class supports.
   */
  public static final String[] formats = { CdmConstants.CDM_EXTENSION };

  /**
   * The maximum json semantic version supported by this ObjectModel version.
   */
  public static final String jsonSemanticVersion = CdmDocumentDefinition.getCurrentJsonSchemaSemanticVersion();

  public static CdmDocumentDefinition fromObject(final CdmCorpusContext ctx, final String name, final String nameSpace,
                                               final String path, final DocumentContent obj) {
    final CdmDocumentDefinition doc = ctx.getCorpus().makeObject(CdmObjectType.DocumentDef);
    doc.setFolderPath(path);
    doc.setNamespace(nameSpace);

    if (!Strings.isNullOrEmpty(obj.getSchema())) {
      doc.setSchema(obj.getSchema());
    }

    if (!Strings.isNullOrEmpty(obj.getDocumentVersion())) {
      doc.setDocumentVersion(obj.getDocumentVersion());
    }

    if (obj.getImports() != null) {
      for (final Import imp: obj.getImports()) {
        doc.getImports().add(ImportPersistence.fromData(ctx, imp));
      }
    }

    if (obj.getDefinitions() != null) {
      for (int i = 0; i < obj.getDefinitions().size(); i++) {
        final CdmCollection<CdmObjectDefinition> definitions = doc.getDefinitions();
        final JsonNode d = obj.getDefinitions().get(i);
        if (d.has("dataTypeName")) {
          definitions.add(DataTypePersistence.fromData(ctx, JMapper.MAP.convertValue(d, DataType.class)));
        } else if (d.has("purposeName")) {
          definitions.add(PurposePersistence.fromData(ctx, d));
        } else if (d.has("attributeGroupName")) {
          definitions.add(AttributeGroupPersistence.fromData(ctx, d));
        } else if (d.has("traitName")) {
          definitions.add(TraitPersistence.fromData(ctx, d));
        } else if (d.has("entityShape")) {
          definitions.add(ConstantEntityPersistence.fromData(ctx, d));
        } else if (d.has("entityName")) {
          definitions.add(EntityPersistence.fromData(ctx, d));
        }
      }
    }

    boolean isResolvedDoc = false;
    if (doc.getDefinitions().getCount() == 1 && doc.getDefinitions().get(0).getObjectType() == CdmObjectType.EntityDef) {
      CdmEntityDefinition entity = (CdmEntityDefinition) doc.getDefinitions().get(0);
      CdmTraitReference resolvedTrait = entity.getExhibitsTraits().item("has.entitySchemaAbstractionLevel");
      // Tries to figure out if the document is in resolved form by looking for the schema abstraction trait
      // or the presence of the attribute context.
      isResolvedDoc = resolvedTrait != null && "resolved".equals(resolvedTrait.getArguments().get(0).getValue());
      isResolvedDoc = isResolvedDoc || entity.getAttributeContext() != null;
    }

    if (!Strings.isNullOrEmpty(obj.getJsonSchemaSemanticVersion())) {
      doc.setJsonSchemaSemanticVersion(obj.getJsonSchemaSemanticVersion());
      if (compareJsonSemanticVersion(ctx, doc.getJsonSchemaSemanticVersion()) > 0) {
          String message = "This ObjectModel version supports json semantic version " + jsonSemanticVersion + " at maximum.";
          message += " Trying to load a document with version " + doc.getJsonSchemaSemanticVersion() + ".";
          if (isResolvedDoc) {
            Logger.warning(ctx, tag, "fromObject", doc.getAtCorpusPath(), CdmLogCode.WarnPersistUnsupportedJsonSemVer);
          } else {
            Logger.error(ctx, tag, "fromObject", doc.getAtCorpusPath(), CdmLogCode.ErrPersistUnsupportedJsonSemVer, jsonSemanticVersion, doc.getJsonSchemaSemanticVersion());
          }
      }
    } else {
        Logger.warning(ctx, tag, "fromObject", doc.getAtCorpusPath(), CdmLogCode.WarnPersistJsonSemVerMandatory);
    }

    return doc;
  }
  
  public static CdmDocumentDefinition fromData(CdmCorpusContext ctx, String docName, String jsonData, CdmFolderDefinition folder) {
    try {
      DocumentContent obj = JMapper.MAP.readValue(jsonData, DocumentContent.class);
      return fromObject(ctx, docName, folder.getNamespace(), folder.getFolderPath(), obj);
    } catch (final Exception e) {
      Logger.error(ctx, tag, "fromData", null, CdmLogCode.ErrPersistDocConversionFailure, docName, e.getLocalizedMessage());
      return null;
    }
  }

  public static DocumentContent toData(final CdmDocumentDefinition instance, final ResolveOptions resOpt,
                                       final CopyOptions options) {
    final DocumentContent documentContent = new DocumentContent();

    documentContent.setSchema(instance.getSchema());
    documentContent.setJsonSchemaSemanticVersion(instance.getJsonSchemaSemanticVersion());
    documentContent.setImports(Utils.listCopyDataAsCdmObject(instance.getImports(), resOpt, options));
    documentContent.setDefinitions(Utils.listCopyDataAsCdmObject(instance.getDefinitions(), resOpt, options));
    documentContent.setDocumentVersion(instance.getDocumentVersion());

    return documentContent;
  }

  /**
   * Compares the document version with the json semantic version supported.
   * 1 => if documentSemanticVersion > jsonSemanticVersion
   * 0 => if documentSemanticVersion == jsonSemanticVersion or if documentSemanticVersion is invalid
   * -1 => if documentSemanticVersion < jsonSemanticVersion
   */
  private static int compareJsonSemanticVersion(CdmCorpusContext ctx, String documentSemanticVersion) {
      String[] docSemanticVersionSplit = documentSemanticVersion.split("\\.");
      String[] currSemanticVersionSplit = jsonSemanticVersion.split("\\.");

      if (docSemanticVersionSplit.length != 3) {
        Logger.warning(ctx, tag, "compareJsonSemanticVersion", null, CdmLogCode.WarnPersistJsonSemVerInvalidFormat);
        return 0;
      }

      for (int i = 0; i < 3; ++i) {
          if (!docSemanticVersionSplit[i].equals(currSemanticVersionSplit[i])) {
              try {
                int version = Integer.parseInt(docSemanticVersionSplit[i]);
                return  version < Integer.parseInt(currSemanticVersionSplit[i]) ? -1 : 1;
              } catch (NumberFormatException e) {
                Logger.warning(ctx, tag, "compareJsonSemanticVersion", null, CdmLogCode.WarnPersistJsonSemVerInvalidFormat);
                return 0;
              }
          }
      }

      return 0;
  }
}
